scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala

```diff
@@ -492,7 +498,12 @@ private[spel] class Typer(
       childElementType: TypingResult
   ) = {
     val isSingleElementSelection = List("$", "^").map(node.toStringAST.startsWith(_)).foldLeft(false)(_ || _)
-    if (isSingleElementSelection) childElementType else parentType
+
+    if (isSingleElementSelection)
+      childElementType
+    else
+      // TODO: limitation - parentType has to lose known value, as properly determining it would require evaluating the expression
```

Comment:
don't really understand this
1. What will the consequences of this be? Why cant it work as before?
2. Is there a test that shows this changed behavior?
3. If its a todo then what should be done to do this todo?

Reply:
1. before this wasn't an issue with Lists, as they always had `valueOpt=None`, while for Maps (records) it also didn't work properly before, but there wasn't a test that showed this.
Perfect behaviours should pass this test:

```
  test("detect proper selection types - map") {
    typeExpression("{'field1': 1, 'field2': 2}.?[(#this.value==1)]").validValue.finalResult.typingResult shouldBe
      Typed.record(
        Map(
          "field1" -> TypedObjectWithValue(Typed.typedClass[Int], 1)
        )
      )
  }
```

Old behaviour resulted in 
```
      Typed.record(
        Map(
          "field1" -> TypedObjectWithValue(Typed.typedClass[Int], 1),
          "field2" -> TypedObjectWithValue(Typed.typedClass[Int], 2)
        )
      )
```

While new behaviours results in 
` Typed.record(Map.empty )`

3. not sure to be honest, perhaps nothing, I'm open to suggestions
still, I think the changes aren't a downgrade as they are